File: Classes/CarTableCell.m
// CarTableCell.m
//
// Kraftstoff


#import "CarTableCell.h"
#import "FuelCalculatorController.h"
#import "AppDelegate.h"
#import <CoreText/CoreText.h>
#import "kraftstoff-Swift.h"

// Standard cell geometry
static CGFloat const PickerViewCellWidth        = 290.0;
static CGFloat const PickerViewCellHeight       =  44.0;
static CGFloat const PickerViewCellMargin       =  10.0;
static CGFloat const PickerViewCellTextPosition =  13.0;

static NSInteger maximumDescriptionLength = 24;


// Attributes for custom PickerViews
static NSDictionary *prefixAttributesDict       = nil;
static NSDictionary *suffixAttributesDict       = nil;


@implementation CarTableCell

+ (void)initialize
{
    CFStringRef font = CFSTR ("Helvetica");

    if (prefixAttributesDict == nil) {

        CTFontRef helvetica24 = CTFontCreateWithName (font, 24, NULL);

        prefixAttributesDict = @{(NSString *)kCTFontAttributeName:(__bridge id)helvetica24,
                                 (NSString *)kCTForegroundColorAttributeName:(id)[[UIColor blackColor] CGColor]};

        CFRelease (helvetica24);
    }

    if (suffixAttributesDict == nil) {

        CTFontRef helvetica18 = CTFontCreateWithName (font, 18, NULL);

        suffixAttributesDict = @{(NSString *)kCTFontAttributeName:(__bridge id)helvetica18,
                                 (NSString *)kCTForegroundColorAttributeName:(id)[[UIColor darkGrayColor] CGColor]};

        CFRelease (helvetica18);
    }
}


- (void)finishConstruction
{
	[super finishConstruction];

    self.carPicker = [[UIPickerView alloc] init];

    self.carPicker.showsSelectionIndicator = YES;
    self.carPicker.dataSource              = self;
    self.carPicker.delegate                = self;

    self.textField.inputView = self.carPicker;
}


- (void)prepareForReuse
{
    [super prepareForReuse];

	self.fetchedObjects = nil;
    [self.carPicker reloadAllComponents];
}


- (void)configureForData:(id)dataObject viewController:(id)viewController tableView:(UITableView *)tableView indexPath:(NSIndexPath *)indexPath
{
	[super configureForData:dataObject viewController:viewController tableView:tableView indexPath:indexPath];

    // Array of possible cars
    self.fetchedObjects = ((NSDictionary *)dataObject)[@"fetchedObjects"];

    // Look for index of selected car
    NSManagedObject *managedObject = [self.delegate valueForIdentifier:self.valueIdentifier];
    NSUInteger initialIndex = [self.fetchedObjects indexOfObject:managedObject];

    if (initialIndex == NSNotFound)
        initialIndex = 0;

    // (Re-)configure car picker and select the initial item
    [self.carPicker reloadAllComponents];
    [self.carPicker selectRow:initialIndex inComponent:0 animated:NO];

    [self selectCar:self.fetchedObjects[initialIndex]];
}


- (void)selectCar:(NSManagedObject *)managedObject
{
    // Update textfield in cell
    NSString *description = [NSString stringWithFormat:@"%@ %@",
                                [managedObject valueForKey:@"name"],
                                [managedObject valueForKey:@"numberPlate"]];

    if ([description length] > maximumDescriptionLength)
        description = [NSString stringWithFormat:@"%@%C",
                        [description substringToIndex:maximumDescriptionLength],
                        (unsigned short)0x2026];

    self.textFieldProxy.text = description;

    // Store selected car in delegate
    [self.delegate valueChanged:managedObject identifier:self.valueIdentifier];
}



#pragma mark -
#pragma mark UIPickerViewDataSource



- (NSInteger)numberOfComponentsInPickerView:(UIPickerView *)pickerView
{
    return 1;
}


- (NSInteger)pickerView:(UIPickerView *)pickerView numberOfRowsInComponent:(NSInteger)component
{
    return [self.fetchedObjects count];
}


- (void)pickerView:(UIPickerView *)pickerView didSelectRow:(NSInteger)row inComponent:(NSInteger)component
{
    [self selectCar:self.fetchedObjects[row]];
}



#pragma mark -
#pragma mark UIPickerViewDelegate



- (CGFloat)pickerView:(UIPickerView *)pickerView rowHeightForComponent:(NSInteger)component
{
    return PickerViewCellHeight;
}


- (CGFloat)pickerView:(UIPickerView *)pickerView widthForComponent:(NSInteger)component
{
    return PickerViewCellWidth;
}


- (CTLineRef)truncatedLineForName:(NSString *)name info:(NSString *)info
{
    NSAttributedString *truncationString = [[NSAttributedString alloc]
                                                initWithString:[NSString stringWithFormat:@"%C", (unsigned short)0x2026]
                                                    attributes:suffixAttributesDict];

    NSMutableAttributedString *attributedString = [[NSMutableAttributedString alloc]
                                                        initWithString:[NSString stringWithFormat:@"%@  %@", name, info]
                                                            attributes:suffixAttributesDict];

    [attributedString setAttributes:prefixAttributesDict range:NSMakeRange (0, [name length])];

    CTLineRef line            = CTLineCreateWithAttributedString ((__bridge CFAttributedStringRef) attributedString);
    CTLineRef truncationToken = CTLineCreateWithAttributedString ((__bridge CFAttributedStringRef) truncationString);
    CTLineRef truncatedLine   = CTLineCreateTruncatedLine (line, PickerViewCellWidth - 2*PickerViewCellMargin, kCTLineTruncationEnd, truncationToken);

    CFRelease (line);
    CFRelease (truncationToken);


    return truncatedLine;
}


- (UIView *)pickerView:(UIPickerView *)pickerView viewForRow:(NSInteger)row forComponent:(NSInteger)component reusingView:(UIView *)view
{
    // Strings to be displayed
    NSManagedObject *managedObject = self.fetchedObjects[row];
    NSString *name = [managedObject valueForKey:@"name"];
    NSString *info = [managedObject valueForKey:@"numberPlate"];


    // Draw strings with attributes into image
    UIImage *image;

    UIGraphicsBeginImageContextWithOptions (CGSizeMake (PickerViewCellWidth, PickerViewCellHeight), NO, 0.0);
    {
        CGContextRef context = UIGraphicsGetCurrentContext();

        CGContextTranslateCTM (context, 1, PickerViewCellHeight);
        CGContextScaleCTM (context, 1, -1);

        CTLineRef truncatedLine = [self truncatedLineForName:name info:info];
        CGContextSetTextPosition (context, PickerViewCellMargin, PickerViewCellTextPosition);
        CTLineDraw (truncatedLine, context);
        CFRelease (truncatedLine);

        image = UIGraphicsGetImageFromCurrentImageContext();
    }
    UIGraphicsEndImageContext();


    // Wrap with imageview
    PickerImageView *imageView;

    if (view != nil && [view isKindOfClass:[PickerImageView class]]) {

        imageView       = (PickerImageView*)view;
        imageView.image = image;

	} else {
        imageView = [[PickerImageView alloc] initWithImage:image];
	}

    imageView.userInteractionEnabled = YES;
    imageView.pickerView = pickerView;
    imageView.rowIndex   = row;


    // Description for accessibility
	imageView.isAccessibilityElement = YES;
    imageView.accessibilityLabel = [NSString stringWithFormat:@"%@ %@", name, info];

    return imageView;
}

@end

File: Classes/CarTableCell.swift
//
//  CarTableCell.swift
//  kraftstoff
//
//  Created by Ingmar Stein on 03.05.15.
//
//

import UIKit

final class CarTableCell: EditableProxyPageCell, UIPickerViewDataSource, UIPickerViewDelegate {
  private var carPicker: UIPickerView
  var cars: [Car]

  // Standard cell geometry
  private let pickerViewCellWidth: CGFloat = 290.0
  private let pickerViewCellHeight: CGFloat = 44.0

  // Attributes for custom PickerViews
  private var prefixAttributes: [NSAttributedString.Key: Any] = [.font: UIFont.preferredFont(forTextStyle: UIFont.TextStyle.headline),
                                                                 .foregroundColor: UIColor.black]
  private var suffixAttributes: [NSAttributedString.Key: Any] = [.font: UIFont.preferredFont(forTextStyle: UIFont.TextStyle.subheadline),
                                                                 .foregroundColor: UIColor.suffix]

  required init() {
    carPicker = UIPickerView()
    cars = []

    super.init()

    let carPickerHeightConstraint = NSLayoutConstraint(item: carPicker, attribute: .height, relatedBy: .equal, toItem: nil, attribute: .notAnAttribute, multiplier: 1.0, constant: 162.0)
    carPickerHeightConstraint.priority = UILayoutPriority(rawValue: 750)
    carPicker.dataSource = self
    carPicker.delegate = self
    carPicker.translatesAutoresizingMaskIntoConstraints = false
    carPicker.isHidden = true

    let stackView = UIStackView(arrangedSubviews: [carPicker])
    stackView.translatesAutoresizingMaskIntoConstraints = false
    stackView.axis = .vertical
    stackView.alignment = .center

    contentView.addSubview(stackView)

    let constraints1 = [carPickerHeightConstraint]
    let constraints = constraints1
      + NSLayoutConstraint.constraints(withVisualFormat: "|-[stackView]-|", options: [], metrics: nil, views: ["stackView": stackView])
      + NSLayoutConstraint.constraints(withVisualFormat: "V:[keyLabel]-[stackView]|", options: [], metrics: nil, views: ["keyLabel": keyLabel, "stackView": stackView])
    NSLayoutConstraint.activate(constraints)
  }

  @available(*, unavailable)
  required init(coder _: NSCoder) {
    fatalError("init(coder:) has not been implemented")
  }

  override func prepareForReuse() {
    super.prepareForReuse()

    cars = []
    carPicker.reloadAllComponents()
  }

  override func configureForData(_ dictionary: [String: Any], viewController: UIViewController, tableView: UITableView, indexPath: IndexPath) {
    super.configureForData(dictionary, viewController: viewController, tableView: tableView, indexPath: indexPath)

    // Array of possible cars
    cars = dictionary["fetchedObjects"] as? [Car] ?? []

    // Look for index of selected car
    guard let car = delegate.valueForIdentifier(valueIdentifier) as? Car else { return }
    let initialIndex = cars.firstIndex(of: car) ?? 0

    // (Re-)configure car picker and select the initial item
    carPicker.reloadAllComponents()
    carPicker.selectRow(initialIndex, inComponent: 0, animated: false)

    selectCar(cars[initialIndex])
  }

  private func selectCar(_ car: Car) {
    // Update textfield in cell
    textFieldProxy.text = "\(car.ksName) \(car.ksNumberPlate)"

    // Store selected car in delegate
    delegate.valueChanged(car, identifier: valueIdentifier)
  }

  private func showPicker(_ show: Bool) {
    carPicker.isHidden = !show
  }

  // MARK: - UIPickerViewDataSource

  func numberOfComponents(in _: UIPickerView) -> Int {
    1
  }

  func pickerView(_: UIPickerView, numberOfRowsInComponent _: Int) -> Int {
    cars.count
  }

  func pickerView(_: UIPickerView, didSelectRow row: Int, inComponent _: Int) {
    selectCar(cars[row])
  }

  // MARK: - UIPickerViewDelegate

  func pickerView(_: UIPickerView, rowHeightForComponent _: Int) -> CGFloat {
    pickerViewCellHeight
  }

  func pickerView(_: UIPickerView, widthForComponent _: Int) -> CGFloat {
    pickerViewCellWidth
  }

  func pickerView(_: UIPickerView, viewForRow row: Int, forComponent _: Int, reusing view: UIView?) -> UIView {
    // Strings to be displayed
    let car = cars[row]
    let name = car.ksName
    let info = car.ksNumberPlate

    var label: UILabel! = view as? UILabel
    if label == nil {
      label = UILabel(frame: .zero)
      label.lineBreakMode = .byTruncatingTail
    }

    let attributedText = NSMutableAttributedString(string: "\(name)  \(info)", attributes: suffixAttributes)
    attributedText.beginEditing()
    attributedText.setAttributes(prefixAttributes, range: NSRange(location: 0, length: name.count))
    attributedText.endEditing()
    label.attributedText = attributedText

    // Description for accessibility
    label.isAccessibilityElement = true
    label.accessibilityLabel = "\(name) \(info)"

    return label
  }

  // MARK: - UITextFieldDelegate

  func textFieldDidBeginEditing(_: UITextField) {
    showPicker(true)
  }

  override func textFieldDidEndEditing(_: UITextField, reason _: UITextField.DidEndEditingReason) {
    showPicker(false)
  }
}
